### PR TITLE
[Desktop][Linux] Add RUNPATH $ORIGIN to flutter_linux_gtk

### PR DESCRIPTION
**Pull request by dcharkes**
When using `dart:ffi`'s `DynamicLibrary.open(...)`, the `RUNPATH` of `bundle/lib/libflutter_linux_gtk.so` is used rather than `bundle/flutter_app`. `libflutter_linux_gtk.so` did not have a `RUNPATH` set, making `dlopen` only succeed on absolute paths and dynamically linked libraries. Setting `RUNPATH` enables us to _dynamically load_ shared libraries without _dynamically linking_ them upfront.

With this fix, the `${plugin}_bundled_libraries` can work without modifying a flutter app:

`${plugin}/linux/CMakeLists.txt`

```cmake
# List of absolute paths to libraries that should be bundled with the plugin
set(mylib_dylib_bundled_libraries
  "${CMAKE_CURRENT_SOURCE_DIR}/../native/lib/linux_x64/libmylib_dylib.so;${CMAKE_CURRENT_SOURCE_DIR}/../native/lib/linux_x64/libmylib_dylib_dependency.so"
  PARENT_SCOPE
)
```

Before this fix, we also needed to modify a `flutter_app` to dynamically link the libraries:

`${flutter_app}/linux/CMakeLists.txt`

```cmake
# Link bundled libraries from plugins into main executable to enable dlopen
# with only library name.
foreach(plugin ${FLUTTER_PLUGIN_LIST})
  if(${plugin}_bundled_libraries)
    target_link_libraries(
      ${BINARY_NAME}
      PUBLIC
      ${${plugin}_bundled_libraries}
    )
  endif()
endforeach(plugin)
```

This `flutter_app` modification is obsolete after this PR.

Tested locally with this patch and `flutter --local-engine=host_release --local-engine-src-path=$HOME/flt/engine/src run linux`.
Change can be observed with `readelf -d flutter_app/build/linux/x64/release/bundle/lib/libflutter_linux_gtk.so | grep RUNPATH`.

Relevant issues: https://github.com/flutter/flutter/issues/78819#issuecomment-912581009 and https://github.com/dart-lang/ffi/issues/118.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: shell/platform/linux/BUILD.gn [hunk 1]
```diff
@@ -220,6 +220,11 @@ executable("flutter_linux_unittests") {
 shared_library("flutter_linux_gtk") {
   deps = [ ":flutter_linux" ]
 
+  if (is_linux) {
```
Comment (stuartmorgan) [inline shell/platform/linux/BUILD.gn:223]: Why are you checking for Linux in a Linux target?

File: shell/platform/linux/BUILD.gn [hunk 2]
```diff
@@ -220,6 +220,11 @@ executable("flutter_linux_unittests") {
...
+    # Ensure we can dlopen bundled dlls in bundle/lib on Flutter Desktop.
```
Comment (stuartmorgan) [inline shell/platform/linux/BUILD.gn:224]: The engine shouldn't reference bundling details that are controlled by the default runner template. Someone could use this library and not even use the template.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
/cc @cbracken 

It seems reasonable to have a default here, but shouldn't we also override this to `$ORIGIN/lib/` in the install step in the template so that just loading the library by name works?

**Comment by dcharkes** [comment]
Thanks for the quick review!

> It seems reasonable to have a default here, but shouldn't we also override this to `$ORIGIN/lib/` in the install step in the template so that just loading the library by name works?

Using `$ORIGIN/lib/` doesn't work. `libflutter_linux_gtk.so` is already in `lib/`, so looking up things from native code invoked by `dart:ffi` or `DynamicLibrary.open` would search in `lib/lib/`. (Or am I missing something?)

The `install` step in the app does not modify the rpath in CMake, it simply copies the files:

```dart
install(FILES "${FLUTTER_LIBRARY}" DESTINATION "${INSTALL_BUNDLE_LIB_DIR}"
  COMPONENT Runtime)
```

I'd need to do some more digging on how to teach CMake to change the `RUNPATH` of a precompiled library (rather than one it is compiling itself).

**Comment by stuartmorgan** [comment]
> Using `$ORIGIN/lib/` doesn't work. `libflutter_linux_gtk.so` is already in `lib/`, so looking up things from native code invoked by `dart:ffi` or `DynamicLibrary.open` would search in `lib/lib/`. (Or am I missing something?)

Isn't `$ORIGIN` always the executable? I'm having trouble finding much documentation, but it sounds like that's its behavior.

> The `install` step in the app does not modify the rpath in CMake, it simply copies the files

I've swapped this out from when I did all the CMake setup, but IIRC CMake automatically does rpath adjustment during the install phase. IIRC that's part of why I used it to do the bundling.

**Comment by dcharkes** [comment]
> Isn't `$ORIGIN` always the executable? I'm having trouble finding much documentation, but it sounds like that's its behavior.

Apparently it is the dynamic library. (At least that is my conclusion after experimenting with this and seeing `$ORIGIN/lib` fail in the dynamic library. I also could not find a lot of documentation.)

> I've swapped this out from when I did all the CMake setup, but IIRC CMake automatically does rpath adjustment during the install phase. IIRC that's part of why I used it to do the bundling.

CMake only does RPATH for non-IMPORTED libraries/executables as far as I can see. And we're building this library in GN/ninja and there we don't have an install step afaik.

**Review by stuartmorgan — APPROVED** [review]
> > Isn't `$ORIGIN` always the executable? I'm having trouble finding much documentation, but it sounds like that's its behavior.
> 
> Apparently it is the dynamic library. (At least that is my conclusion after experimenting with this and seeing `$ORIGIN/lib` fail in the dynamic library. I also could not find a lot of documentation.)

Interesting. Well, that renders the CMake question moot, so LGTM.

**Pull request merged**